aggregators/merger.go

```diff
@@ -395,27 +397,49 @@ func mergeHistogram(to, from *aggregationpb.HDRHistogram) {
 		return
 	}
 
-	requiredLen := len(to.Buckets) + len(from.Buckets)
-	for toIdx, fromIdx := 0, 0; toIdx < len(to.Buckets) && fromIdx < len(from.Buckets); {
-		v := to.Buckets[toIdx] - from.Buckets[fromIdx]
-		switch {
-		case v == 0:
-			// For every bucket that is common, we need one less bucket in final slice
-			requiredLen--
-			toIdx++
-			fromIdx++
-		case v < 0:
-			toIdx++
-		case v > 0:
-			fromIdx++
+	var extra int
+	toLen, fromLen := len(to.Buckets), len(from.Buckets)
+	if fromLen < toLen { // Heuristics to trade between O(m lg n) and O(n + m).
+		// Fast path to optimize for cases where len(from.Buckets) << len(to.Buckets)
+		// Binary search for all from.Buckets in to.Buckets for fewer comparisons,
+		// mergeHistogram will be O(m lg n) where m = fromLen and n = toLen.
+		for fromIdx := 0; fromIdx < fromLen; fromIdx++ {
+			toIdx, found := sort.Find(toLen, func(toIdx int) int {
+				return int(from.Buckets[fromIdx] - to.Buckets[toIdx])
+			})
+			if found {
+				to.Counts[toIdx] += from.Counts[fromIdx]
+				from.Counts[fromIdx] = 0
+			} else {
+				extra++
+			}
+		}
```

Comment:
oh this is clever

Reply:
Actually I have an idea that will be applicable to more cases. Instead of searching for 0..toLen each time, we utilize the `toIdx` from the previous pass. Will require us to invert the loop `for fromIdx := 0; fromIdx < fromLen; fromIdx++ {`.

Reply:
Moreover, in your code we must ensure extra > 0 when we break, and we'll also have to correctly determine `extra`. These shortcircuits are dangerous :cry:  
```
			if toIdx == len(findIn) {
				break
			}
```

Reply:
Updated the PR to limit the search space with the help of result of the previous iteration.

Reply:
> if any of the bucket differs we fallback to previous logic

Yes we can fallback to the previous logic, but we'll still have to pick between a O(n+m) or O(m lg n) counting to grow the slice properly. Since in a lot of cases m << n, having O(m lg n) just for counting may still be faster than O(n+m). So it isn't that clear cut. Am I missing anything?